### PR TITLE
L2-364: Neuron start dissolving

### PR DESCRIPTION
**Pull request by lmuntanerDfinity**
# Motivation

User can start and stop dissolving process in a neuron.

# Changes

* Neuron services: isNeuronControllable, startDissolving and stopDissolving.
* Governance api: startDissolving and stopDissolving.
* Upgrade nns-js.
* Change button to DissolveActionButton that handles both cases.
* Account services: getAccountByPrincipal

# Tests

* New tests for the neuron services
* New tests for the governance api
* Test for the new button.
* New test for the account services.

### CODE REVIEW

**Comment by lmuntanerDfinity** [comment]
UX in the video:

https://www.loom.com/share/67b1ade91a564dd98dbec779cbbce936

New Confirmation modal:

Screenshot 2022-03-30 at 14 51 40

File: frontend/svelte/src/lib/services/accounts.services.ts [hunk 1]
```diff
@@ -44,3 +46,19 @@ export const addSubAccount = async ({
...
+ * Returns the principal's main or hardware account
+ *
+ * Subaccounts do not have Principal
+ */
+export const getAccountByPrincipal = (
+  principal: string
+): Account | undefined => {
+  const accounts = get(accountsStore);
+
+  if (accounts.main?.principal?.toText() === principal) {
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/services/accounts.services.ts:60]: The comment says "principal's main or hardware account" vs the code mention "main" only.

Should we add a TODO or should we correct the comment or else?

Comment (peterpeterparker) [inline frontend/svelte/src/lib/services/accounts.services.ts:60]: isn't a `?` missing, `accounts?...`?

Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/services/accounts.services.ts:60]: Yes, good point. Do we have the jira issue to add the hardware wallets? I haven't found it.

File: frontend/svelte/src/lib/services/neurons.services.ts [hunk 2]
```diff
@@ -389,3 +408,18 @@ export const loadNeuron = ({
...
+ * Returns true if the neuron can be controlled. A neuron can be controlled if:
+ *
+ *  1. The user is the controller
+ *  OR
+ *  2. The user's hardware wallet is the controller.
+ *
+ */
+// TODO: Ask: what's the difference between this and property `isCurrentUserController` in the fullNeuron
+// TODO: Ask: Which actions are disabled by which scenarios
+export const isNeuronControllable = (neuron: NeuronInfo): boolean =>
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/services/neurons.services.ts:422]: nth: `= ({fullNeuron}: NeuronInfo):`

File: frontend/svelte/src/lib/types/account.ts [hunk 3]
```diff
@@ -1,8 +1,11 @@
 import type { ICP } from "@dfinity/nns";
+import type { Principal } from "@dfinity/principal";
 import type { SubAccountArray } from "../canisters/nns-dapp/nns-dapp.types";
 
 export interface Account {
   identifier: string;
+  // Main account has Principal
+  principal?: Principal;
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/types/account.ts:8]: out of curiosity, is the main account principal equals to the identity principal or the two are different?

**Review by peterpeterparker — CHANGES_REQUESTED** [review]
Just small questions

File: frontend/svelte/src/lib/services/accounts.services.ts (hunk 1 above)
Comment (peterpeterparker) [reply frontend/svelte/src/lib/services/accounts.services.ts:60]: Indeed. You can use L2-433 for example.

File: frontend/svelte/src/lib/services/accounts.services.ts [hunk 4]
```diff
@@ -44,3 +46,19 @@ export const addSubAccount = async ({
...
+    return accounts.main;
+  }
+  return;
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/services/accounts.services.ts:63]: isn't it the same? I don't mind being explicit though;

File: frontend/svelte/src/lib/services/accounts.services.ts (hunk 1 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/services/accounts.services.ts:60]: Nops:
```
  const initialAccounts: AccountsStore = {
    main: undefined,
    subAccounts: undefined,
  };
```

File: frontend/svelte/src/lib/services/accounts.services.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterpeterparker) [reply frontend/svelte/src/lib/services/accounts.services.ts:63]: In js yes but in ts feels like being used to return `void`

File: frontend/svelte/src/lib/services/accounts.services.ts (hunk 1 above)
Comment (peterpeterparker) [reply frontend/svelte/src/lib/services/accounts.services.ts:60]: Sorry my bad, thanks for the answer

**Comment by lmuntanerDfinity** [comment]
@peterpeterparker ready for another review

**Review by peterpeterparker — APPROVED** [review]
LGTM

Being said, not against a review of @mstrasinskis before merge if you don't mind

**Review by mstrasinskis — APPROVED** [review]
LGTM, just one question?

File: frontend/svelte/src/lib/services/neurons.services.ts [hunk 5]
```diff
@@ -166,17 +169,13 @@ const checkNeuronBalances = async (neurons: NeuronInfo[]): Promise<void> => {
   return listNeurons({ skipCheck: true });
 };
 
-export const updateDelay = async ({
+const getAndLoadNeuronHelper = async ({
```
Comment (mstrasinskis) [inline frontend/svelte/src/lib/services/neurons.services.ts:172]: Out of curiosity: what does the word `helper` means in this context?

Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/services/neurons.services.ts:172]: It makes sense to have a couple of tests and it's not so hard to add it.

File: frontend/svelte/src/lib/services/accounts.services.ts [hunk 6]
```diff
@@ -44,3 +46,21 @@ export const addSubAccount = async ({
 
   await syncAccounts();
 };
+
+/*
+ * Returns the principal's main or hardware account
+ *
+ * Subaccounts do not have Principal
+ */
+export const getAccountByPrincipal = (
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/services/accounts.services.ts:55]: It makes sense to have a couple of tests and it's not so hard to add it.

**Pull request merged**